### PR TITLE
Bbs updates to draft 3

### PR DESCRIPTION
**Pull request by dwaite**
Update BBS text to draft 3 operations as well as update code to generate examples.

Note that I am still experimenting with the best approach for line wrapping the generated examples, so there may be subsequent changes in the generation code.

### CODE REVIEW

File: draft-ietf-jose-json-proof-algorithms.md [hunk 1]
```diff
@@ -219,256 +219,79 @@ See the example in the appendix of the JSON Web Proof draft.
...
-### BLS Curve
-
-The pairing friendly elliptic curve used for the BBS software implementation is part of the BLS family with an embedding degree of 12 over a 381-bit prime field.  For this JPA, only the group G2 is used.
+### Algorithm
 
-In the implementation the method used to generate the key pairs is `generateBls12381G2KeyPair()`.
+The `BBS-DRAFT-3` "alg" parameter value corresponds to a ciphersuite identifier of `BBS_BLS12381G1_XMD:SHA-256_SSWU_RO_H2G_HM2S_` from [@!I-D.irtf-cfrg-bbs-signatures], used in a JWP issuance form, while `BBS-PROOF-DRAFT-3` corresponds to the same ciphersuite in presentation form.
 
-### Messages
+### Key format
```
Comment (selfissued) [inline draft-ietf-jose-json-proof-algorithms.md:228]:
```suggestion
### Key Format
```

File: draft-ietf-jose-json-proof-algorithms.md [hunk 2]
```diff
@@ -219,256 +219,79 @@ See the example in the appendix of the JSON Web Proof draft.
...
 
-BBS is a multi-message scheme and operates on an array of individual messages for signing and proof generation.  Each message is a single binary octet string.  The BBS implementation uses a hash-to-curve method to map each message to a point.
+The key used for the `BBS-DRAFT-3` algorithm is an elliptic curve-based key pair, specifically against the G_2 subgroup of a pairing friendly curve. Additional details on key generation can be found in [@!I-D.irtf-cfrg-bbs-signatures, section 3.3]
```
Comment (selfissued) [inline draft-ietf-jose-json-proof-algorithms.md:230]:
```suggestion
The key used for the `BBS-DRAFT-3` algorithm is an elliptic curve-based key pair, specifically against the G_2 subgroup of a pairing friendly curve. Additional details on key generation can be found in [@!I-D.irtf-cfrg-bbs-signatures, Section 3.3]
```

File: draft-ietf-jose-json-proof-algorithms.md [hunk 3]
```diff
@@ -219,256 +219,79 @@ See the example in the appendix of the JSON Web Proof draft.
...
-### Verification
+<{{./fixtures/build/bbs-issuer.json.jwp}}
+Figure: Issued JWP (JSON serialization)
 
-The verifier decodes the JWP issuer protected header and payload values into a messages array, skipping any non-revealed payloads.  The current BBS implementation embeds the revealed indices into the output proof value, so the verification messages array only needs to include the disclosed messages.
+<{{./fixtures/build/bbs-issuer.compact.jwp}}
+Figure: Issued JWP (compact serialization)
 
-In the implementation, the method used to verify the proof is `blsVerifyProof({verifyProof, publicKey, [issuer_header, payload2, ...], presentation_header)`.
+### Issuance proof verification
```
Comment (selfissued) [inline draft-ietf-jose-json-proof-algorithms.md:259]:
```suggestion
### Issuance Proof Verification
```

File: draft-ietf-jose-json-proof-algorithms.md [hunk 4]
```diff
@@ -219,256 +219,79 @@ See the example in the appendix of the JSON Web Proof draft.
...
-Figure: bbs-issuer-protected-header
-
-The first payload is the string `"Doe"` with the octet sequence of `[ 34, 68, 111, 101, 34 ]` and base64url-encoded as `IkRvZSI`.
+### Presentation
 
-The second payload is the string `"Jay"` with the octet sequence of `[ 34, 74, 97, 121, 34 ]` and base64url-encoded as `IkpheSI`.
+Derivation of a presentation is done by the holder using the `ProofGen` operation from [@!I-D.irtf-cfrg-bbs-signatures, section 3.4.3].
 
-The third payload is the string `"jaydoe@example.org"` with the octet sequence of `[ 34, 106, 97, 121, 100, 111, 101, 64, 101, 120, 97, 109, 112, 108, 101, 46, 111, 114, 103, 34 ]` and base64url-encoded as `ImpheWRvZUBleGFtcGxlLm9yZyI`.
+This operation utilizes the issuer's public key as `PK`, the issuer protected header as `header`, the issuance proof as `signature`, the issuance payloads as `messages`, and the holder's presentation protected header as `ph`.
```
Comment (selfissued) [inline draft-ietf-jose-json-proof-algorithms.md:269]: It seems inconsistent that `PK` is capitalized and `ph` is not.  Is there a reason that we shouldn't capitalize `PH`?  Or are these identifiers from other specs, in which case we should use them as-is?

File: draft-ietf-jose-json-proof-algorithms.md [hunk 5]
```diff
@@ -219,256 +219,79 @@ See the example in the appendix of the JSON Web Proof draft.
...
-    "NDI"
-  ],
-  "proof": "tANC_gnNFFivUloiGrJQ4VvReBe5n0xJvexzjR9TKyq698TsRhN7UPmS
-  7aww0ME-ZDuaFjSluPpHNGrpGvD71nqFPfFGf1PwcIK1l6DWK9VT0-6_AUGHk-LFGG
-  i3CY3PFWqIoXOOA8SbNK7N1A2u3A"
-}
-```
-Figure: bbs-issued-jwp
+<{{./fixtures/template/bbs-prover-presentation-header.json}}
+Figure: Holder presentation header
```
Comment (selfissued) [inline draft-ietf-jose-json-proof-algorithms.md:280]:
```suggestion
Figure: Holder Presentation Header
```

File: draft-ietf-jose-json-proof-algorithms.md [hunk 6]
```diff
@@ -219,256 +219,79 @@ See the example in the appendix of the JSON Web Proof draft.
...
-19, 172, 217, 65, 63, 246, 39, 6, 30, 77, 132, 86, 36, 41, 3, 234,
-72, 146, 200, 101, 150, 159, 108, 140, 15, 195, 57, 249, 154, 191,
-204, 91, 30, 159, 32, 157, 24, 3, 110, 90, 102, 99, 206, 42, 58, 1,
-181, 215, 85, 29, 32, 131, 46, 76, 25, 5, 43, 203, 32, 215, 167,
-169, 108, 56, 174, 146, 51, 174, 40, 190, 22, 37, 93, 156, 245,
-208, 26, 55, 180, 135, 115, 70, 96, 106, 243, 213, 131, 196, 63,
-165, 42, 157, 22, 94, 46]
-```
-Figure: bbs-present-proof
+### Presentation verification
```
Comment (selfissued) [inline draft-ietf-jose-json-proof-algorithms.md:288]:
```suggestion
### Presentation Verification
```

File: draft-ietf-jose-json-proof-algorithms.md [hunk 7]
```diff
@@ -219,256 +219,79 @@ See the example in the appendix of the JSON Web Proof draft.
...
-  fausG99w6GKAAAAAIFgnhW_xwhkRSVwwgEyNSyQ5PmrsAJnl6zkD88Uv_YBFVs0cLR
-  sWpF1-uxU_QBw2aHYxR5B_wau84Q-oYB_8VcgmnxryMW0mWecdbeAwSovPsi1dPglp
-  MmpOWX4t-8tbTM5DprN-iUtMcqtX876eq8AAAABF3EHyaXaecu5C5WxIjUr6pTFU4T
-  4NN6B1xHEatCejiCLROs2UE_9icGHk2EViQpA-pIkshllp9sjA_DOfmav8xbHp8gnR
-  gDblpmY84qOgG111UdIIMuTBkFK8sg16epbDiukjOuKL4WJV2c9dAaN7SHc0ZgavPV
-  g8Q_pSqdFl4u"
-}
-```
-Figure: bbs-present-jwp
+Verification of a presentation is done by the holder using the `ProofVerify` operation from [@!I-D.irtf-cfrg-bbs-signatures, section 3.4.4].
```
Comment (selfissued) [inline draft-ietf-jose-json-proof-algorithms.md:290]:
```suggestion
Verification of a presentation is done by the holder using the `ProofVerify` operation from [@!I-D.irtf-cfrg-bbs-signatures, Section 3.4.4].
```

File: draft-ietf-jose-json-proof-algorithms.md [hunk 8]
```diff
@@ -219,256 +219,79 @@ See the example in the appendix of the JSON Web Proof draft.
...
-wY7hUihoC6hJl7Vlic9ssX_fjZ6QGo0AzB3o5aHN_xgqSQUwAPpRmckfMnWPn2uZgD3E
-i0AAAAdIW0OjVpeHzjoE7lStFvpGW3VnrUfloX5G24SUtysY6yWWtkvbtKj6fausG99w
-6GKAAAAAIFgnhW_xwhkRSVwwgEyNSyQ5PmrsAJnl6zkD88Uv_YBFVs0cLRsWpF1-uxU_
-QBw2aHYxR5B_wau84Q-oYB_8VcgmnxryMW0mWecdbeAwSovPsi1dPglpMmpOWX4t-8tb
-TM5DprN-iUtMcqtX876eq8AAAABF3EHyaXaecu5C5WxIjUr6pTFU4T4NN6B1xHEatCej
-iCLROs2UE_9icGHk2EViQpA-pIkshllp9sjA_DOfmav8xbHp8gnRgDblpmY84qOgG111
-UdIIMuTBkFK8sg16epbDiukjOuKL4WJV2c9dAaN7SHc0ZgavPVg8Q_pSqdFl4u
-```
-Figure: bbs-present-compact
+This operation utilizes the issuer's public key as `PK`, the issuer protected header as `header`, the issuance proof as `signature`, the holder's presentation protected header as `ph`, and the payloads as `disclosed_messages`.
```
Comment (selfissued) [inline draft-ietf-jose-json-proof-algorithms.md:292]: Same comment about the inconsistent capitalization of PK and ph from above.

**Review by selfissued — APPROVED** [review]
Looks good.  Thanks for doing this!  Please accept my editorial suggestions before merging.

Also, please add an acknowledgement saying that the BBS examples were generated using https://github.com/mattrglobal/pairing_crypto .

File: draft-ietf-jose-json-proof-algorithms.md (hunk 4 above)
Comment (dwaite) [reply draft-ietf-jose-json-proof-algorithms.md:269]: That is the capitalization within the BBS Signatures spec

File: draft-ietf-jose-json-proof-algorithms.md (hunk 8 above)
Comment (dwaite) [reply draft-ietf-jose-json-proof-algorithms.md:292]: see above - aim was to capitalize these the same as the BBS spec

File: draft-ietf-jose-json-proof-algorithms.md (hunk 4 above)
Comment (selfissued) [reply draft-ietf-jose-json-proof-algorithms.md:269]: OK - then leave it that way.  Thanks.

**Pull request merged**